DM-44803 - Implement new function to collect Test Cases of a Test Cycle from Zephyr Scale

b1quint opened the pull request:
Implement new function to collect Test Cases of a Test Cycle from Zephyr Scale.

b1quint committed:
Update .gitignore with vscode folder

b1quint committed:
Update ``zephyr_interface.py`` with a new class to hold Test Cycle methods

b1quint committed:
Add test for test authentication with JIRA API

b1quint committed:
Add some unit tests to run without CI and connecting with JIRA

b1quint committed:
Add test case for ZephyrScale to run in CI

b1quint committed:
Add methods to parse fields from json payload

b1quint committed:
In zephyr_scale.py, remove TestCycle and run example

sebastian-aranda commented on tests/test_zephyr_interface_no_ci.py:
```diff
@@ -0,0 +1,220 @@
+            list(test_executions["values"][0].keys()), values_expected_keys
+        )
+
+    async def test_parse_environment_from_id(self):
+
+        environment_id = 6824992  # Daytime environment id for '1. Daytime'
+        environment = await self.zapi.parse_environment_from_id(environment_id)
+        self.assertEqual(environment, "1. Daytime")
+
+    async def test_parse_project_from_id(self):
+
+        project_id = 350001  # BLOCK project id
+        project = await self.zapi.parse_project_from_id(project_id)
+        self.assertEqual(project, "BLOCK")
+
+    async def test_parse_status_from_id(self):
+
+        status_id = 3940035  # Pass status id
+        status = await self.zapi.parse_status_from_id(status_id)
+        self.assertEqual(status, "Pass")
```
I suggest to also add tests for:
- `parse_test_case_from_id`
- `parse_test_cycle_from_id`

sebastian-aranda replied:
Another suggestion that could be too much, but could be good for better maintenance: move all constants such as `environment_id`, `project_id`, `status_id` and even others as e.g. `test_cycle_key` (in other tests) to a config file (e.g. `tests/config.py` or `tests/constants.py`).

b1quint replied:
When writing these tests, I found out that there is no API interface that I can use to write `parse_test_case_from_id` . So I am removing any reference to it.

b1quint replied:
When writing these tests, I found out that there is no API interface that I can use to write `parse_test_case_from_id` . So I am removing any reference to it.

b1quint replied:
Regarding your second comment, I created global variables at the beginning of `tests/test_zephyr_interface_no_ci.py`. Not as pretty as you suggested, but it is a quick-and-dirty solution.

sebastian-aranda replied:
having global variables at the beginning sounds great too! thanks for looking into this.

sebastian-aranda commented on tests/test_zephyr_interface_no_ci.py:
I really liked the `no_ci` approach! Even do this won't run on CI as one would expect, we can still use it from time to time to make sure the payloads of the Jira/Zephyr API are still the ones we expect. I recommend to add some details to the `README.md` to explain the purpose of these tests and a couple lines of instructions about how to run them.

sebastian-aranda replied:
Also, for the sake of been more explicit I recommend to also test missing nested parameters such as: `project` field (or others) that come with the `self` and `id` keys, this way we can also easily identify in the tests which are the parameters that need to be "parsed".

sebastian-aranda replied:
Finally, is there any reason because you didn't test the `get_test_cycle` function here? If not, I recommend to do it.

sebastian-aranda commented on tests/test_zephyr_interface_no_ci.py:
```diff
@@ -0,0 +1,220 @@
+        self.assertEqual(test_case_name, expected_test_case_name)
+        self.assertEqual(test_case_version, expected_test_case_version)
+
+    async def test_get_test_executions(self):
```
Just for the sake of better code ordering, I recommend to move this test above `test_extract_test_case_from_test_execution`.

b1quint replied:
Hmmm... in which sense? I tried to have the methods inside the test class organized alphabetically.

sebastian-aranda replied:
Sorry, dismiss this comment. At first glance I thought this test should have been placed above because its structure is similar to the first tests above (like defining the payload and then calling some functions), however ordering alphabetically is good! Sorry for the noise :pray:.

sebastian-aranda commented on tests/test_zephyr_interface_no_ci.py:
```diff
@@ -0,0 +1,220 @@
+        self.assertEqual(test_case["key"], test_case_key)
+        self.assertListEqual(list(test_case.keys()), payload_expected_keys)
+
+    async def test_get_test_case_steps(self):
```
This test and all the bellow (in this same file) are missing the `@pytest.mark.asyncio` fixture.

sebastian-aranda commented on python/lsst/ts/planning/tool/zephyr_interface.py:
Why did you remove the `get_test_cycles` function? It was part of the required methods for the frontend UI. If there is a good reason I recommend to explain that on the PR description, and perhaps instead of removing it, just commenting the code and adding a comment to explain that will be rolled back later.

b1quint replied:
Fixed. I put it back.

sebastian-aranda replied:
thanks!

b1quint replied:
I think I know why I removed the `get_test_cycles` function. There is a challenge in writing unit tests for the tests that involve the payload. Ideally, we want a payload with values already parsed. Right now, the `get_test_cycles` is the only method that parses values automatically. 

I started working a solution in PR#4. Have a look [here](https://github.com/lsst-ts/ts_planning_tool/blob/c3519126306994ca6e27ba34dd91ae05880815ba/python/lsst/ts/planning/tool/zephyr_interface.py#L210) to have an idea of what I am talking about. 

In order to close this PR, I propose that `get_test_cycle` does not parse anything yet. And we deal with that in PR#4. What do you think?

sebastian-aranda replied:
I see, thanks for the clarification! This sounds good to me :+1:.

sebastian-aranda reviewed (changes requested):
Great job @b1quint, this is looking good! I left several comments which I think should be good to incorporate before merging the PR.

b1quint committed:
Put back get_test_cycles method

b1quint committed:
In parse_test_case_from_id, add NonImplemented error

Strangely, Zephyr Scale does not support parsing test cases from ID

b1quint committed:
Update unit tests based on comments in PR#3
- Real data is now represented as global variables
- Moved test_parse_test_cycle_from_id up
- Add missing marks to unit tests

b1quint committed:
Add unit test for get_test_cycle plus minor fixes

sebastian-aranda reviewed (approved):
Looks good to me :+1: , thank you for the good work and taking care of the comments @b1quint !

b1quint committed:
Update README.md with instructions to run unit tests
